### PR TITLE
feat: `dtm develop create-plugin` implement

### LINKED ISSUE #325: proposal: create a cmd tool to help generate scaffolding for plugin code
## Description

If we need to develop a new plugin, there will be a lot of `repetitive work`. There is a lot of repetitive code. Copy-pasting of this code is tedious and error-prone. So I want to develop a tool to do this.

## Describe the Proposal

cmd eg.

`dtm plugin create --name=your-plugin-name --type=your-plugin-type`

The first version will be: `dtm plugin create --name=you-plugin-name`. But it may be necessary to support the below commands in the future:

- `dtm plugin rename --name=your-plugin-name --new-name=your-new-plugin-name`
- `dtm plugin delete --name=your-plugin-name`
- `dtm plugin create --name=your-plugin-name --type=helm`

---

- I wanna put the template file (tpl) in the `main repo` stream so it can be run offline. And the update will not need to modify multiple repos at the same time. This update will be more frequent than repo-scaffolding plugins, because this structure is closely related to the `dtm` itself, so I think it will be easier to put the tpl stay with `stream`.
- When this `create` command is executed, it will put many of the generated files in the corresponding directory, and then leave a `TODO: xxx` comment in it.
- I'll create a new package `internal/pkg/pluginbuilder` and put all logic inside this pkg(Of course, the entry cmd will still be in the `cmd package`. But inside cmd is just very simple "code").

### PR DESCRIPTION
**Pull request by daniel-hutao**
Signed-off-by: Daniel Hu <tao.hu@merico.dev>

# Summary

feat: `dtm develop create-plugin` implement

## Description

This pr only contains code implementation, and the corresponding usage documentation is left in a separate pr, perhaps on the blog we will run recently.

- `./dtm-darwin-arm64 develop create-plugin --name="foo"`

```sh
2022-03-18 13:23:20 ℹ [INFO]  Render template files finished.
2022-03-18 13:23:20 ℹ [INFO]  Persist all files finished.

The DevStream PMC (project management committee) sincerely thank you for your devotion and enthusiasm in creating new plugins!

To make the process easy as a breeze, DevStream(dtm) has generated some templated source code files for you to flatten the learning curve and reduce manual copy-paste.
In the generated templates, dtm has left some special marks in the format of "TODO(dtm)".
Please look for these TODOs by global search. Once you find them, you will know what to do with them. Also, please remember to check our documentation on creating a new plugin:

**README_when_create_plugin.md**

Source code files created.

Happy hacking, buddy!
Please give us feedback through GitHub issues if you encounter any difficulties. We guarantee that you will receive unrivaled help from our passionate community!

```

The files generated:

image

## Related Issues

close #325 

## Last but not least

Thanks to @IronCore864 for helping me reorganize the last output log, this modification makes this feature more shiny!

### CODE REVIEW

**Comment by daniel-hutao** [comment]
@warren830 @imxw @summingyu 
You can help to review this pr together, if you are interested and have time. It's interesting!

File: cmd/devstream/develop.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func developCMDFunc(cmd *cobra.Command, args []string) {
+	if err := validateArgs(args); err != nil {
+		log.Fatal(err)
+	}
+
+	developAction := develop.Action(args[0])
+	log.Debugf("The develop action is: %s", developAction)
+	if err := develop.BranchAction(developAction); err != nil {
```
Comment (IronCore864) [inline cmd/devstream/develop.go:30]: I think we can have a better name here for "Branch Action". I understand here the purpose is to switch on the action, but the name "branch action" sounds weird. Maybe "ExecuteAction(developAction)" is good enough?

File: internal/pkg/develop/develop.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/merico-dev/stream/pkg/util/log"
+)
+
+type Action string
+
+const (
+	ActionCreatePlugin Action = "create-plugin"
+)
+
+var ActionSet = map[Action]struct{}{
```
Comment (IronCore864) [inline internal/pkg/develop/develop.go:14]: Is "ActionSet" needed to be exported?

File: internal/pkg/develop/plugin/plugin.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	return buf.String(), nil
+}
+
+// PersistFiles gets the []pluginTpl.File, for each File:
+// call the persistFile() method to deal with.
+func (p *Plugin) PersistFiles(files []pluginTpl.File) error {
+	fileCount := len(files)
+	log.Debugf("There are %d files wait to persist.", fileCount)
+	for i, file := range files {
+		log.Debugf("Persist process: %d/%d", i+1, fileCount)
```
Comment (IronCore864) [inline internal/pkg/develop/plugin/plugin.go:99]: Missing "period" (.) at the end of the log sentence. Please check for all log lines; but in general it's very good, almost perfect logging. Kudos!

File: internal/pkg/develop/plugin/create.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	name := viper.GetString("name")
+	if name == "" {
+		return fmt.Errorf("the name must be not \"\", you can specify it by --name flag")
+	}
+	log.Debugf("Got the name: %s.", name)
+
+	p := NewPlugin(name)
+
+	// 1. Render template files
+	files, err := p.RenderTplFiles()
```
Comment (IronCore864) [inline internal/pkg/develop/plugin/create.go:25]: It seems "RenderTplFiles" comes from the same package. Do we have to export it? Same for "PersistFiles".

File: internal/pkg/develop/plugin/plugin.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		log.Debugf("Failed to write content to the file: %s.", err)
+	}
+	log.Debugf("File %s has been created.", filePath)
+
+	return nil
+}
+
+func (p *Plugin) PrintHelpInfo() {
+	help := `
+The DevStream PMC (project management committee) sincerely thank you for your devotion and enthusiasm in creating new plugins!
```
Comment (IronCore864) [inline internal/pkg/develop/plugin/plugin.go:132]: Fantastic message :)

**Review by IronCore864 — CHANGES_REQUESTED** [review]

File: internal/pkg/develop/develop.go (hunk 2 above)
Comment (daniel-hutao) [reply internal/pkg/develop/develop.go:14]: Yes, it will be use at `cmd/develop` package

File: internal/pkg/develop/plugin/plugin.go (hunk 3 above)
Comment (daniel-hutao) [reply internal/pkg/develop/plugin/plugin.go:99]: Good point. I changed some logs and added some periods before committing. I thought about checking all the logs, but I still forgot. I'll take a check now.

File: internal/pkg/develop/plugin/create.go (hunk 4 above)
Comment (daniel-hutao) [reply internal/pkg/develop/plugin/create.go:25]: Yes, these 2 functions are only used in the same package now.  But I still choose to `Export` it for these 2 are the main functions at the `Create()` logic. I think it's more clear to write:
```go
func (p *Plugin) RenderTplFiles(){}
func (p *Plugin) renderTplFile(){}
```

compare to

```go
func (p *Plugin) renderTplFiles(){}
func (p *Plugin) renderTplFile(){}
```

File: internal/pkg/develop/plugin/plugin.go (hunk 5 above)
Comment (daniel-hutao) [reply internal/pkg/develop/plugin/plugin.go:132]: +1
Kudos to us!

File: cmd/devstream/develop.go (hunk 1 above)
Comment (daniel-hutao) [reply cmd/devstream/develop.go:30]: agree

**Review by lfbdev — COMMENTED** [review]
It looks like the modification to the Makefile is missing.

insert:

go build -buildmode=plugin -trimpath -gcflags="all=-N -l" -o .devstream/NAME-${GOOS}-${GOARCH}_${VERSION}.so ./cmd/NAME/

**Review by lfbdev — COMMENTED** [review]
Also modification in build_linux_amd64.sh:

go build -buildmode=plugin -trimpath -gcflags="all=-N -l" -o output/NAME-${GOOS}-${GOARCH}_${VERSION}.so ./cmd/NAME/

**Comment by daniel-hutao** [comment]
> It looks like the modification to the Makefile is missing.
> 
> insert:
> 
> go build -buildmode=plugin -trimpath -gcflags="all=-N -l" -o .devstream/NAME-${GOOS}-${GOARCH}_${VERSION}.so ./cmd/NAME/

@lfbdev See the `internal/pkg/develop/plugin/template/README_when_create_plugin.md.go` for more info.

**Review by lfbdev — APPROVED** [review]

**Review by IronCore864 — APPROVED** [review]

**Pull request merged**